### PR TITLE
don't support cmake 3.12, 3.13, 3.14

### LINKED ISSUE #24983: [编译错误] cmake版本为3.12和3.17，PaddlePaddle源码编译报nvcc错误
- 标题：Linux编译错误
- 版本、环境信息：
    1）PaddlePaddle版本：8603b5fb722692bfee148ddf30ee9c3a7d39a5e1
    2）CPU：MKL/OpenBlas/MKLDNN/ 都是使用默认编译选项
    3）GPU：V100，CUDA Version 9.0.176
    4）系统环境：Ubuntu 16.04.4 LTS, Python2.7  
- 复现信息：
    1）cmake命令：
    `cmake .. -DWITH_GPU=ON -DWITH_TESTING=ON -DCMAKE_BUILD_TYPE=Release -DPY_VERSION=2.7`
    2)  cmake为3.10版本，编译正常。cmake为3.12和3.17版本，编译中报nvcc错误，如下
    `nvcc fatal : redefinition of argument 'std'` 。

> 内部汇报case，可以联系 juncai

### PR DESCRIPTION
**Pull request by T8T9**
### PR types

Bug fixes

### PR changes

Others

### Describe

fix #24983

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -13,6 +13,12 @@
 # limitations under the License
 
 cmake_minimum_required(VERSION 3.10)
+
+if((${CMAKE_VERSION} VERSION_GREATER_EQUAL 3.12) AND (${CMAKE_VERSION} VERSION_LESS 3.15))
+    message(FATAL_ERROR "cmake ${CMAKE_VERSION} is not supported because of bug https://tinyurl.com/ydbb9m7n. "
+       "cmake 3.16 is recommended. how to install: https://cmake.org/install/")
```
Comment (luotao1) [inline CMakeLists.txt]: 1. You can use cmake 3.16 (recommended), 3.10, 3.11 or 3.17. Please refer to the install document: https://cmake.org/install/
2. 加一下对于`WITH_GPU`的限制

Comment (T8T9) [reply CMakeLists.txt]: done

File: CMakeLists.txt [hunk 2]
```diff
@@ -27,6 +27,13 @@ include(init)
 # TODO(Shibo Tao): remove find_package(CUDA) completely.
 find_package(CUDA QUIET)
 option(WITH_GPU         "Compile PaddlePaddle with NVIDIA GPU"          ${CUDA_FOUND})
+
+# cmake 3.12, 3.13, 3.14 will append gcc link options to nvcc, and nvcc doesn't recognize them.
+if(WITH_GPU AND (${CMAKE_VERSION} VERSION_GREATER_EQUAL 3.12) AND (${CMAKE_VERSION} VERSION_LESS 3.15))
+    message(FATAL_ERROR "cmake ${CMAKE_VERSION} is not supported when WITH_GPU=ON because of bug https://tinyurl.com/ydbb9m7n. "
+       "You can use cmake 3.16 (recommended), 3.10, 3.11 or 3.17. Please refer to the install document: https://cmake.org/install/")
```
Comment (luotao1) [inline CMakeLists.txt]: 刚评论里漏了3.15了，可以补上。

Comment (T8T9) [reply CMakeLists.txt]: done

File: CMakeLists.txt [hunk 3]
```diff
@@ -27,6 +27,13 @@ include(init)
 # TODO(Shibo Tao): remove find_package(CUDA) completely.
 find_package(CUDA QUIET)
 option(WITH_GPU         "Compile PaddlePaddle with NVIDIA GPU"          ${CUDA_FOUND})
+
+# cmake 3.12, 3.13, 3.14 will append gcc link options to nvcc, and nvcc doesn't recognize them.
+if(WITH_GPU AND (${CMAKE_VERSION} VERSION_GREATER_EQUAL 3.12) AND (${CMAKE_VERSION} VERSION_LESS 3.15))
+    message(FATAL_ERROR "cmake ${CMAKE_VERSION} is not supported when WITH_GPU=ON because of bug https://tinyurl.com/ydbb9m7n. "
```
Comment (luotao1) [inline CMakeLists.txt:33]: > https://tinyurl.com/ydbb9m7n

该链接改成：https://cmake.org/pipermail/cmake/2018-September/068195.html 会更直观

Comment (T8T9) [reply CMakeLists.txt:33]: done, https://github.com/PaddlePaddle/Paddle/pull/25130/

**Pull request merged**